### PR TITLE
stricter date parsing

### PR DESCRIPTION
**Pull request by SimonCropp**
Not sure on this one. DateGraphType, will always return  "yyyy-MM-dd" but it will (previously) accept any date format  as show by the failing tests. eg `2015-12-01T14:15:07.123Z`. So it seems weird to me that someone can use the same graph type and pushing a string up will come back as a diff format and with less resolution (data loss)

### CODE REVIEW

**Comment by SimonCropp** [comment]
i should also point out that the current date parsing seems to be susceptible to ambiguous dates.  eg `09-10-2015")`. which i didnt think should be allowed under ISO_8601?

**Comment by SimonCropp** [comment]
@joemcbride thoughts on this one. If we r going to change it we should do it as part of the major

**Comment by SimonCropp** [comment]
@joemcbride gentle bump

**Comment by SimonCropp** [comment]
@sungam3r thoughts?

**Comment by sungam3r** [comment]
This is one of those issues that I was particularly interested in but still can not find the time to look. I will try to read in the near future.

File: src/GraphQL/Types/DateGraphType.cs [hunk 1]
```diff
@@ -17,13 +18,27 @@ public override object Serialize(object value)
...
 
             return null;
         }
 
-        public override object ParseValue(object value) => ValueConverter.ConvertTo(value, typeof(DateTime));
+        public override object ParseValue(object value)
+        {
+            if (value is DateTime dateTime)
+            {
+                return dateTime;
```
Comment (sungam3r) [inline src/GraphQL/Types/DateGraphType.cs]: Wouldn’t it be right to take only the part responsible for the date? Or it’s generally worth generating an error if the format of the input data does not match the required format for `Date`. I believe it is wrong to allow the use of a date with a time component for a date scalar

File: src/GraphQL/Types/DateGraphType.cs [hunk 2]
```diff
@@ -17,13 +18,27 @@ public override object Serialize(object value)
...
+            }
+
+            var valueAsString = (string)value;
```
Comment (sungam3r) [inline src/GraphQL/Types/DateGraphType.cs]: consider to change this to `if (value is string valueAsString)`

File: src/GraphQL/Types/DateGraphType.cs (hunk 1 above)
Comment (SimonCropp) [reply src/GraphQL/Types/DateGraphType.cs]: fixed to throw an exception

File: src/GraphQL/Types/DateGraphType.cs (hunk 2 above)
Comment (SimonCropp) [reply src/GraphQL/Types/DateGraphType.cs]: yep. agreed. done

File: src/GraphQL/Types/DateGraphType.cs [hunk 3]
```diff
@@ -17,13 +18,34 @@ public override object Serialize(object value)
...
+            if (value is string valueAsString)
+            {
+                if (DateTime.TryParseExact(valueAsString, "yyyy-MM-dd", DateTimeFormatInfo.InvariantInfo, DateTimeStyles.None, out var date))
+                {
+                    return date;
+                }
+                throw new FormatException($"Could not parse date. Expected yyyy-MM-dd. Value: {valueAsString}");
+            }
+
+            throw new FormatException($"Could not parse date. Expected either a string or a DateTime. Value: {value}");
```
Comment (sungam3r) [inline src/GraphQL/Types/DateGraphType.cs]:
```suggestion
            throw new FormatException($"Could not parse date. Expected either a string or a DateTime without time component. Value: {value}");
```

**Comment by sungam3r** [comment]
So you still need to fix the tests that work with the date.

File: src/GraphQL.Tests/Types/DateGraphTypeTests.cs [hunk 4]
```diff
@@ -20,52 +20,12 @@ public void serialize_string_to_date()
             });
         }
 
-        [Fact]
-        public void serialize_local_date_returns_date_only()
```
Comment (sungam3r) [inline src/GraphQL.Tests/Types/DateGraphTypeTests.cs]: Don't you think that we should to keep these tests, just now expect an exception from them about time component?

**Comment by sungam3r** [comment]
@SimonCropp Will you continue to work on this PR? I would like these changes to be merged.

**Comment by sungam3r** [comment]
@SimonCropp friendly bump

**Comment by SimonCropp** [comment]
@sungam3r thanks for the reminder. this should be good now. but it will need a squash on merge

**Comment by sungam3r** [comment]
@Shane32 This will need to be noted in the documentation as breaking change for Data scalar type.

**Comment by Shane32** [comment]
Sure, I'll add it to the upgrade notes.  Correct me if I'm wrong, but looks like now `yyyy-MM-dd` is the only valid format when deserializing, and when serializing, the `DateTime` must not have a time component.

**Comment by sungam3r** [comment]
Right.

**Comment by SimonCropp** [comment]
yep. at my current employer we have been running with a custom date type for over a year now. will be good to remove it

**Comment by nbarbettini** [comment]
> Sure, I'll add it to the upgrade notes. Correct me if I'm wrong, but looks like now `yyyy-MM-dd` is the only valid format when deserializing, and when serializing, the `DateTime` must not have a time component.

@Shane32 Friendly reminder to include this in your upgrade PR: https://github.com/graphql-dotnet/graphql-dotnet/pull/1798/
It caused a bit of head scratching for me during a 2.4->3.0 upgrade.

**Comment by Shane32** [comment]
Thanks for the reminder @nbarbettini !

**Comment by sungam3r** [comment]
> It caused a bit of head scratching for me during a 2.4->3.0 upgrade.

I also had to explain possible problems to our teams. Problems did arise.

**Pull request merged**